esp-hal-common/src/analog/adc/cal_curve.rs

```diff
@@ -0,0 +1,250 @@
+use core::marker::PhantomData;
+
+use crate::adc::{
+    AdcCalEfuse,
+    AdcCalLine,
+    AdcCalScheme,
+    AdcHasLineCal,
+    Attenuation,
+    RegisterAccess,
+};
+
+const COEFF_MUL: i64 = 1 << 52;
+
+type CurveCoeff = i64;
+
+/// Polynomial coefficients for specified attenuation.
+pub struct CurveCoeffs {
+    /// Attenuation
+    atten: Attenuation,
+    /// Polynomial coefficients
+    coeff: &'static [CurveCoeff],
+}
+
+type CurvesCoeffs = &'static [CurveCoeffs];
+
+/// Marker trait for ADC which support curve futting
+///
+/// See also [`AdcCalCurve`].
+pub trait AdcHasCurveCal {
+    /// Coefficients for calculating the reading voltage error.
+    ///
+    /// A sets of coefficients for each attenuation.
+    const CURVES_COEFFS: CurvesCoeffs;
+}
+
+/// Curve fitting ADC calibration scheme
+///
+/// This scheme implements final polynomial error correction using predefined
+/// coefficient sets for each attenuation.
+///
+/// This scheme also includes basic calibration ([`AdcCalBasic`]) and line
+/// fitting ([`AdcCalLine`]).
+#[derive(Clone, Copy)]
+pub struct AdcCalCurve<ADCI> {
+    line: AdcCalLine<ADCI>,
+
+    /// Coeff of each term. See `adc_error_coef_atten` for details (and the
+    /// magic number 2)
+    coeff: &'static [CurveCoeff],
+
+    _phantom: PhantomData<ADCI>,
+}
+
+impl<ADCI> AdcCalCurve<ADCI> {
+    fn calc_error(&self, val: u16) -> i32 {
+        if val == 0 {
```

Comment:
I don't think this shortcut is very useful.

Reply:
Hmm I might not have been clear with this comment. What I means was the `== 0 { // early return` part, as it's probably not a very common case that the ADC measures an exact 0. My apologies.

It's a bit hard to see what happened to this function, because you force pushed. Did you inline `calc_error` into `adc_val`, or do I just misremember how this was organized before?